src/runtime.rs

```diff
@@ -0,0 +1,371 @@
+                })
+                .collect::<Vec<_>>();
+            Runtime {
+                handle: Handle {
+                    sender: Arc::new(Mutex::new(Some(task_sender))),
+                },
+                name,
+                handles,
+            }
+        }
+
+        pub fn handle(&self) -> &Handle {
+            &self.handle
+        }
+
+        /// Allow the given duration for a graceful shutdown, then abandon
+        /// the threads and return.
+        pub fn shutdown_timeout(mut self, duration: Duration) {
+            tracing::trace!("shutting down runtime with timeout");
+            // send the shutdown signal by dropping the sender
+            self.handle.sender.lock().unwrap().take();
+            // wait for all threads to finish
+            let start = std::time::Instant::now();
+            while start.elapsed() < duration {
+                if self.handles.iter().all(|h| h.is_finished()) {
+                    break;
+                }
+                std::thread::sleep(std::time::Duration::from_millis(10));
+            }
+            self.handles.clear();
+        }
+
+        /// Shutdown the runtime entirely in the background
+        pub fn shutdown_background(self) {
+            self.shutdown_timeout(Duration::from_secs(0));
+        }
+    }
+
+    /// Signal to the runtime that it should shutdown, and wait indefinitely
+    /// for it to do so.
+    impl Drop for Runtime {
+        fn drop(&mut self) {
+            tracing::trace!("dropping sender");
+            self.handle.sender.lock().unwrap().take();
+            tracing::trace!("waiting for runtimes to terminate");
+            for handle in self.handles.drain(..) {
+                handle.join().ok();
+            }
+        }
+    }
+
+    impl Handle {
+        fn sender(&self) -> flume::Sender<Task> {
+            let inner = self.sender.lock().unwrap();
+            let sender = inner.as_ref().unwrap().clone();
+            sender
+        }
+
+        pub async fn spawn<F, Fut>(&self, f: F)
+        where
+            F: FnOnce() -> Fut + Send + 'static,
+            Fut: Future + 'static,
+        {
+            let f = || f().map(|_| ()).boxed_local();
+            self.sender()
+                .into_send_async(Task(Box::new(f)))
+                .await
+                .unwrap();
+        }
+
+        pub fn spawn_sync<F, Fut>(&self, f: F)
+        where
+            F: FnOnce() -> Fut + Send + 'static,
+            Fut: Future + 'static,
+            Fut::Output: 'static,
+        {
+            let f = || f().map(|_| ()).boxed_local();
+            self.sender().send(Task(Box::new(f))).unwrap();
+        }
+
+        pub async fn run<F, Fut>(&self, f: F) -> Fut::Output
+        where
+            F: FnOnce() -> Fut + Send + 'static,
+            Fut: Future + 'static,
+            Fut::Output: Send + 'static,
+        {
+            let (tx, rx) = futures::channel::oneshot::channel();
```

Comment:
we already have `flume` as a dependency, so maybe use that for all these, and not mix in channels from futures and std?

Reply:
flume is not an oneshot channel. I assume that a flume channel is more expensive. I can use a tokio::sync::oneshot::channel if you want.